### PR TITLE
Preserve string behavior with unicode characters

### LINKED ISSUE #29055: TorchScript does not preserve string behavior for string with unicode characters
## 🐛 Bug

If a function calls `len(s)`, it works differently in scripted and eager modes.

## To Reproduce

Test:
```
import torch
string = "Normanðy"
print(string)

def foo(s: str):
    return len(s)

foo_scripted = torch.jit.script(foo)

print(foo(string))
print(foo_scripted(string))
```

Output:
```
Normanðy
8
9
```

cc @suo

### PR DESCRIPTION
**Pull request by ansley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47096 Preserve string behavior with unicode characters**

### CODE REVIEW

**Review by ZolotukhinM — COMMENTED** [review]
I we decide to support unicode in TorchScript we might need to look at broader picture: there are more functions that we'd need to support. E.g. we might want to index a unicode string, currently it also fails:
```
import torch

def foo(s: str):
    print(s)
    x = len(s)
    print(x)
    print(s[6]) # << this fails in scripted version
    return x

s = "Normanðy"
scripted = torch.jit.script(foo)
print('Eager:')
foo(s)
print('Scripted:')
scripted(s)
```

We could either audit all string functions and properly implement them as you did for `len`, or we could consider using `std::wstring` instead of `std::string` in TorchScript which would probably define some of these issues out. @suo I vaguely recall we discussed this issue a long time ago, what do you think about it?

**Comment by suo** [comment]
I agree with @ZolotukhinM: it may be confusing to support unicode in one case but fail in all others. I think it's a good idea to support unicode. Several users that work in NLP-related domains have asked for it. But it's quite a bit of work:
1. Operations on strings should work with unicode data. This includes `len` as implemented here, but also indexing, concatenation, any methods defined on strings, etc.
2. If we want to follow Python 3, then strings in TorchScript (and in C++ API for torchscript) need to be utf-8 encoded by default, and would require breaking changes to how string handling works.
3. Like Python, we would probably need to define a `bytes` type, which represents the raw storage for a string (as opposed to its encoded representation). Currently, our `string` type is really `bytes`.
4. If we want to support unicode string literals in our code, or unicode identifiers, the compiler frontend needs to be changed to support unicode as well.
5. Some additional functionality needs to be defined if we want really robust support for unicode strings. For example, it's potentially useful to provide `unicodedata.normalize()` if we want to give people a TorchScript-supported solution to the notorious `"ê" == "ê" -> False` situation.

I don't think we need to do all of these things to have reasonable support for unicode data, but at least 1 and 2 probably are necessary? In the past, we have deferred this work because of resource constraints, but it's definitely worth doing if we have the time.

cc @gmagogsfm @SplitInfinity, as we have discussed this in the past.

**Comment by suo** [comment]
useful reference on unicode handling in python: https://docs.python.org/3/howto/unicode.html

**Comment by ansley** [comment]
Follow up on what @suo and @ZolotukhinM wrote: Do we want to move forward with extended Unicode support? Or even full Unicode support? If we're okay staying on the Basic Multilingual Plane, we could divide up the first task in [Michael's comment](https://github.com/pytorch/pytorch/pull/47096#issuecomment-720656029) into a number of subtasks and use them as bootcamp assignments.

cc @gmagogsfm @SplitInfinity

**Comment by ansley** [comment]
Closing until we have a plan for full Unicode support
